ding/model/template/hpt.py

```diff
@@ -0,0 +1,221 @@
+        token_num = 16
+        self.tokens = nn.Parameter(torch.randn(1, token_num, 128) * INIT_CONST)
+        self.cross_attention = CrossAttention(128, heads=8, dim_head=64, dropout=0.1)
+
+    def compute_latent(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Overview:
+            Compute latent representations of the input data using
+            the feature extractor and cross-attention.
+
+        Arguments:
+            - x (:obj:`torch.Tensor`): Input tensor with shape [B, T, ..., F].
+
+        Returns:
+            - stem_tokens (:obj:`torch.Tensor`): Latent tokens with shape [B, 16, 128].
+        """
+        # Using the Feature Extractor
+        stem_feat = self.feature_extractor(x)
+        stem_feat = stem_feat.reshape(stem_feat.shape[0], -1, stem_feat.shape[-1])  # (B, N, 128)
+        # Calculating latent tokens using CrossAttention
+        stem_tokens = self.tokens.repeat(len(stem_feat), 1, 1)  # (B, 16, 128)
+        stem_tokens = self.cross_attention(stem_tokens, stem_feat)  # (B, 16, 128)
+        return stem_tokens
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Overview:
+            Forward pass to compute latent tokens.
+
+        Arguments:
+            - x (:obj:`torch.Tensor`): Input tensor.
+
+        Returns:
+            - torch.Tensor: Latent tokens tensor.
+        """
+        return self.compute_latent(x)
+
+    def freeze(self):
+        """Freeze the parameters of the model, preventing updates during training."""
+        for param in self.parameters():
+            param.requires_grad = False
+
+    def unfreeze(self):
+        """Unfreeze the parameters of the model, allowing updates during training."""
+        for param in self.parameters():
+            param.requires_grad = True
+
+    def save(self, path: str):
+        """Save the model state dictionary to a file."""
+        torch.save(self.state_dict(), path)
+
+    @property
+    def device(self):
+        """Returns the device on which the model parameters are located."""
+        return next(self.parameters()).device
+
+
+class CrossAttention(nn.Module):
+    """
+    Overview:
+        CrossAttention module used in the Perceiver IO model.
+        It computes the attention between the query and context tensors,
+        and returns the output tensor after applying attention.
+
+    Arguments:
+        query_dim (:obj:`int`): The dimension of the query input.
+        heads (:obj:`int`, optional): The number of attention heads. Defaults to 8.
+        dim_head (:obj:`int`, optional): The dimension of each attention head. Defaults to 64.
+        dropout (:obj:`float`, optional): The dropout probability. Defaults to 0.0.
+    """
+
+    def __init__(self, query_dim: int, heads: int = 8, dim_head: int = 64, dropout: float = 0.0):
+        super().__init__()
+        inner_dim = dim_head * heads
+        context_dim = query_dim
+        self.scale = dim_head ** -0.5
```

Comment:
Why is `self.scale` defined in this manner? Here are the relevant comments to explain its purpose.